Core: Remove configure: INITIALIZE_DEFAULT_CATALOG_FILEIO_FOR_TEST

flyrain opened the pull request:
As we discussed [here](https://github.com/apache/polaris/pull/1532#issuecomment-2868205342), this PR removes the config to make it safer, which means users couldn't turn it on as a configuration.

flyrain commented on service/common/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalog.java:
```diff
@@ -232,7 +222,16 @@ public String name() {
-  public FileIO getIo() {
+  public synchronized FileIO getIo() {
+    if (catalogFileIO == null) {
+      LOGGER.debug(
+          "Initializing a default catalogFileIO with properties {}", tableDefaultProperties);
```
This is a carry-on from the deleted code. We can remove it if people think it's not useful.

XJDKC commented on service/common/src/test/java/org/apache/polaris/service/catalog/io/FileIOFactoryTest.java:
```diff
@@ -198,7 +198,7 @@ public void testLoadFileIOForCleanupTask() {
     // 1. BasePolarisCatalog:doCommit: for writing the table during the creation
     // 2. BasePolarisCatalog:doRefresh: for reading the table during the drop
     // 3. TaskFileIOSupplier:apply: for clean up metadata files and merge files
-    Mockito.verify(testServices.fileIOFactory(), Mockito.times(3))
```
nit: We need to update the comment to reflect the change. The first method call now comes from initializing the default catalog file IO.

snazy reviewed (changes requested):
The config option `INITIALIZE_DEFAULT_CATALOG_FILEIO_FOR_TEST` is still configured in a bunch of places.

The only obvious reason for `IcebergCatalog.catalogFileIO` and `getIo()` is to be a convenience for some test classes. I'd prefer to remove that field and method from `IcebergCatalog` and add the necessary pieces to the test classes that need it. Those can create a FileIO instance and close it after usage.

dimas-b commented on service/common/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalog.java:
```diff
@@ -232,7 +222,16 @@ public String name() {
+    if (catalogFileIO == null) {
+      LOGGER.debug(
+          "Initializing a default catalogFileIO with properties {}", tableDefaultProperties);
+      catalogFileIO = loadFileIO(ioImplClassName, tableDefaultProperties);
```
This used to be performed only when `INITIALIZE_DEFAULT_CATALOG_FILEIO_FOR_TEST` was `true`. Now it appears to be performed in all cases... do I read it correctly? If so, what's the rationale for this behaviour change?

flyrain replied:
`getIo()` and all of its callers are test-only.

dimas-b replied:
Noted. Please make it package-private.

flyrain replied:
Yeah, I actually tried that. Turns out some tests will fail due to across package reference.

dimas-b replied:
But, why not have tests create FileIO instances and inject them into `IcebergCatalog`? I think it would be a lot more clear.

flyrain replied:
What do you mean by "inject"? A setter will face exactly the same issue, which is public and only used by tests. 
Or are you talking about a subclass?

dimas-b replied:
Yes, looking at the test code, `IcebergCatalog` objects are created by tests for tests, so I believe since we started this cleanup effort, it makes sense for these tests to own FileIO creation.

flyrain replied:
The class `IcebergCatalog` belongs to package `org.apache.polaris.service.catalog.iceberg`, while `IcebergCatalogTest` belongs to package `org.apache.polaris.service.quarkus.catalog`. Are you suggesting a setter or a subclass? I think a subclass will make the test code overly complicated. Introducing a setter like `setFileIO(fileIO)` will face the same issue, which is the method is public and used only by tests.  

Please let me know which way you prefer. cc @dimas-b @snazy.

dimas-b replied:
I'd suggest either a test-specific constructor or test-specific `initialize(...)` method, which would then accept a specific FileIO implementation.

flyrain replied:
Added a sub class so that the only the test sub class creates load the fileIO

flyrain commented:
> The config option `INITIALIZE_DEFAULT_CATALOG_FILEIO_FOR_TEST` is still configured in a bunch of places.

Good catch. Removed them in the new commit.

> The only obvious reason for `IcebergCatalog.catalogFileIO` and `getIo()` is to be a convenience for some test classes. I'd prefer to remove that field and method from `IcebergCatalog` and add the necessary pieces to the test classes that need it. Those can create a FileIO instance and close it after usage.

That's fair point. Not only `getIo()`, but also `newTableOps()` and `newViewOps()` are test-only. I added a `@VisibleForTesting` so that they are labeled now. We can move it out in a follow-up PR if needed. WDYT?

snazy commented:
Doing it _cleanly_ here is IMHO better

singhpk234 reviewed (approved):
LGTM !

On orthogonal note, not related to the PR, I would have gone with introducing a generic config like `IS_TESTING` (which spark does [here](https://github.com/apache/spark/blob/master/common/utils/src/main/scala/org/apache/spark/util/SparkEnvUtils.scala#L27)) to make the classes aware, which are doing some logic specifically for testing purpose aware, that its now being executed in Testing mode ([here](https://github.com/apache/polaris/pull/1624#pullrequestreview-2856193006)), rather than introducing a test variable with a name and purpose for this case.
